### PR TITLE
Support format arguments in EXPLAIN ANALYZE

### PR DESCRIPTION
**Pull request by ZacBlanco**
## Description

This change adds two features:

1. Support for `format <X>` in EXPLAIN ANALYZE
2. Adds support for a `stats` field in each plan node representation

This allows the user to get a computer-friendly representation of the plan with the statistics. This can be useful for visualization or as input to other systems for plan node-level performance analysis.

## Motivation and Context

- Additional execution analysis capabilities through query plan JSON

## Impact

- The JSON representation of a plan from now includes a "stats" field representation statistics at runtime. Its contents map to `PlanNodeStats.java`
  - This effects the plans sent from the event listener framework
  - Backwards compatibility should be preserved as the change only introduces a new field in the JSON representation without modifying previously-present fields.

## Test Plan

- New tests in the parser and statement analyzer to verify user-supplied arguments.

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== RELEASE NOTES ==

General Changes
* Improve :doc:`/sql/explain-analyze` statement to support a ``format`` argument with values of ``<TEXT|JSON>`` :pr:`22733`
```

### CODE REVIEW

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

Pull branch, local docs build, looks good. Thanks!

**Comment by steveburnett** [comment]
Suggest revising the release note entry following the [Order of changes](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#order-of-changes) in the Release Notes Guidelines, also add a link to the existing documentation (I tested this link in a local build): 

```
== RELEASE NOTES ==

General Changes
* Improve :doc:`/sql/explain-analyze` statement to support a ``format`` argument with values of ``<TEXT|JSON|GRAPHVIZ>`` :pr:`22733`
```

File: presto-main/src/main/java/com/facebook/presto/operator/ExplainAnalyzeOperator.java [hunk 1]
```diff
@@ -144,8 +155,21 @@ public Page getOutput()
...
+        String plan;
+        switch (format) {
+            case TEXT:
+                plan = textDistributedPlan(queryInfo.getOutputStage().get().getSubStages().get(0), functionAndTypeManager, operatorContext.getSession(), verbose);
+                break;
+            case JSON:
+                plan = jsonDistributedPlan(queryInfo.getOutputStage().get().getSubStages().get(0), functionAndTypeManager, operatorContext.getSession());
+                break;
+            case GRAPHVIZ:
+                plan = graphvizDistributedPlan(queryInfo.getOutputStage().get().getSubStages().get(0), functionAndTypeManager, operatorContext.getSession());
```
Comment (aaneja) [inline presto-main/src/main/java/com/facebook/presto/operator/ExplainAnalyzeOperator.java:167]: Since we are not changing the Graphviz output, should we skip supporting `EXPLAIN ANALYZE (FORMAT GRAPHIZ)` ? (the output is same as `EXPLAIN (FORMAT GRAPHIZ)`)  
IMO, it will only add to confusion for users

Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/ExplainAnalyzeOperator.java:167]: Thanks for your opinion on this one. I had a similar thought but decided to just include it anyways. However, if you think it is confusing I will remove support for it

Comment (aaneja) [reply presto-main/src/main/java/com/facebook/presto/operator/ExplainAnalyzeOperator.java:167]: Yes, lets skip it for now. We can enhance Graphviz output in a different PR (if there's interest) and add it back

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 2]
```diff
@@ -196,7 +197,12 @@ private RelationPlan createExplainAnalyzePlan(Analysis analysis, Explain stateme
         PlanNode root = underlyingPlan.getRoot();
         Scope scope = analysis.getScope(statement);
         VariableReferenceExpression outputVariable = newVariable(variableAllocator, scope.getRelationType().getFieldByIndex(0));
-        root = new ExplainAnalyzeNode(getSourceLocation(statement), idAllocator.getNextId(), root, outputVariable, statement.isVerbose());
+        ExplainFormat.Type type = statement.getOptions()
+                .stream().filter(option -> option instanceof ExplainFormat)
+                .findFirst().map(ExplainFormat.class::cast)
+                .map(ExplainFormat::getType)
+                .orElse(ExplainFormat.Type.TEXT);
```
Comment (ClarenceThreepwood) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:204]:
```suggestion
                .orElse(TEXT);
```

File: presto-main/src/main/java/com/facebook/presto/operator/ExplainAnalyzeOperator.java (hunk 1 above)
Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/ExplainAnalyzeOperator.java:167]: I don't think it's particularly difficult, but I'm not familiar with the graphviz format. It would just require some more time+effort to make it work properly

**Review by ClarenceThreepwood — DISMISSED** [review]

**Review by aaneja — APPROVED** [review]

**Comment by ZacBlanco** [comment]
@feilong-liu @jaystarshot would you like to take a. look at this?

**Review by jaystarshot — APPROVED** [review]

**Comment by feilong-liu** [comment]
> @feilong-liu @jaystarshot would you like to take a look at this?

Will take a look and reply later today

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 3]
```diff
@@ -1135,9 +1138,20 @@ protected Scope visitExplain(Explain node, Optional<Scope> scope)
...
             checkState(node.isAnalyze(), "Non analyze explain should be rewritten to Query");
-            if (node.getOptions().stream().anyMatch(option -> !option.equals(new ExplainType(DISTRIBUTED)))) {
-                throw new SemanticException(NOT_SUPPORTED, node, "EXPLAIN ANALYZE only supports TYPE DISTRIBUTED option");
-            }
+            List<ExplainFormat.Type> formats = node.getOptions().stream()
+                    .filter(option -> option instanceof ExplainFormat)
+                    .map(ExplainFormat.class::cast)
+                    .map(ExplainFormat::getType)
+                    .collect(Collectors.toList());
+            checkState(formats.size() <= 1, "only a single format option is accepted in EXPLAIN ANALYZE");
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1146]: nit:
```suggestion
            checkState(formats.size() <= 1, "only a single format option is supported in EXPLAIN ANALYZE");
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -1135,9 +1138,20 @@ protected Scope visitExplain(Explain node, Optional<Scope> scope)
...
+            formats.stream().findFirst().ifPresent(format -> checkState(format.equals(TEXT) || format.equals(JSON),
+                    "only TEXT and JSON formats are accepted in EXPLAIN ANALYZE"));
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1148]:
```suggestion
                    "only TEXT and JSON formats are supported in EXPLAIN ANALYZE"));
```

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java [hunk 5]
```diff
@@ -846,6 +847,26 @@ public void testExplainAnalyze()
         analyze("EXPLAIN ANALYZE SELECT * FROM t1");
     }
 
+    @Test
+    public void testExplainAnalyzeFormatJson()
+    {
+        analyze("EXPLAIN ANALYZE (format JSON) SELECT * FROM t1");
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:853]: these just check that we don't fail.  is there a way to test/see what the new output looks like?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStats.java [hunk 6]
```diff
@@ -39,41 +42,56 @@ public class PlanNodeStats
 
     private final Duration planNodeScheduledTime;
     private final Duration planNodeCpuTime;
+    private final Duration planNodeBlockedWallTime;
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStats.java:45]: why are these added here?

Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanNodeStats.java:45]: The main driver behind this change is that we have an external tool that helps visualize some plans from the JSON output of the event listener. The event listener output is the actually the same output as what you get from an `EXPLAIN ANALYZE`. However, you only get execution statistics in the text output from the event listener, but not the JSON one. So, to augment the event listener's JSON plan output to be closer to the text version, I added some more statistics to this class.

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java (hunk 5 above)
Comment (ZacBlanco) [reply presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:853]: I added a `deserialize` method on the `JsonRenderedNode` and an associated test inside of `AbstractDistributedTestQueries`

**Review by rschlussel — APPROVED** [review]
